[#162] Show warning when ignoring repeated  (Alternative approach)

Eclipse-Dominator opened the pull request:
Fixes #162 
This PR provides an alternative approach which displays all the fields that are being duplicated as compared to #176 

Currently, repeated occurrences of the same parameter in a command are being ignored without any notification to the user. This can lead to potential user errors going unnoticed, especially when using the wrong parameter prefix multiple times.

This commit introduces a friendly warning system that alerts the user when a parameter is being ignored. Now, if a parameter is repeated in a command, all occurrences except the last one will trigger a warning message, ensuring that the user is aware of any unintentional mistakes or typos.

Let's provide a better user experience by catching and notifying them of ignored parameters, improving the overall usability and preventing potential errors.

Example:
Multiple fields duplicated:
![image](https://github.com/se-edu/addressbook-level3/assets/4567895/d5dbfde3-0e82-4bc2-85ce-ec9e97c2a335)

Single field duplicated:
![image](https://github.com/se-edu/addressbook-level3/assets/4567895/bfeab288-b725-4a9b-8431-794fb485cd3c)

damithc commented on src/main/java/seedu/address/logic/commands/EditCommand.java:
```diff
@@ -49,6 +49,8 @@ public class EditCommand extends Command {
     public static final String MESSAGE_NOT_EDITED = "At least one field to edit must be provided.";
     public static final String MESSAGE_DUPLICATE_PERSON = "This person already exists in the address book.";
+    public static final String MESSAGE_DUPLICATE_FIELDS = "Detected multiple modifications for the following field%s: "
+            + "%s.\nOnly the last change will be applied.\n";
```
`multiple modifications` -> `multiple values`?

damithc replied:
Also, this message uses two words for the same thing `modifications`, `change`.

damithc commented on src/main/java/seedu/address/logic/parser/EditCommandParser.java:
```diff
@@ -44,15 +44,27 @@ public EditCommand parse(String args) throws ParseException {
         if (argMultimap.getValue(PREFIX_PHONE).isPresent()) {
+            if (argMultimap.getAllValues(PREFIX_PHONE).size() > 1) {
+                editPersonDescriptor.addDuplicateField(PREFIX_PHONE.getPrefix());
+            }
```
This `if` block is duplicated a few times. Is there a way to avoid that duplication?

damithc commented on src/test/java/seedu/address/testutil/EditPersonDescriptorBuilder.java:
```diff
@@ -55,6 +68,14 @@ public EditPersonDescriptorBuilder withPhone(String phone) {
+    public EditPersonDescriptorBuilder withDuplicatePhone() {
+        descriptor.addDuplicateField(PREFIX_PHONE.getPrefix());
+        return this;
+    }
```
Instead of adding these methods, can the same be accomplished by calling `withPhone()` multiple times?

Eclipse-Dominator replied:
It's possible but it might be less intuitive to when building the test cases since calling the same method twice may not be that intuitive as compared to calling it via a specific method.

I will try to experiment combining the functionality in the same method.

damithc replied:
I felt calling the method `withPhone()` twice is the more natural way to specify the same parameter twice. With a specific method like `withDuplicatePhone()`, you open the door for calling `withDuplicatePhone()` without calling `withPhone()` first.
Yes, experiment variations and see how it goes.

damithc reviewed (commented):
Added some prelim comments.

damithc commented on src/main/java/seedu/address/logic/parser/EditCommandParser.java:
```diff
@@ -79,4 +79,14 @@ private Optional<Set<Tag>> parseTagsForEdit(Collection<String> tags) throws Pars
+            descriptor.addDuplicateField(prefix.getPrefix());
+        }
+        return argMap.getValue(prefix).isPresent();
+    }
```
Good attempt at reducing duplication, but I'm not sure this method is the way to do it. It is best when a method is either a command (i.e., does something) or a query (i.e., answers a question) but not both. 
For example, someone seeing a call to this method will think it is a query only, thus misleading the reader.

Is there any way to avoid the above flaw, even at the expense of slightly more code?

damithc commented on src/test/java/seedu/address/testutil/EditPersonDescriptorBuilder.java:
```diff
@@ -41,32 +46,40 @@ public EditPersonDescriptorBuilder(Person person) {
      */
     public EditPersonDescriptorBuilder withName(String name) {
+        descriptor.getEmail().ifPresent(x -> descriptor.addDuplicateField(PREFIX_NAME.getPrefix()));
         descriptor.setName(new Name(name));
```
getEmail() -> getName()?
BTW, should the logic in line 52 be inside the setName() method instead?

Eclipse-Dominator replied:
I missed this when copilot generated the code wrongly.

damithc reviewed (commented):
Raised a few ideas about the implementation strategy. Might not work but worth exploring.

Eclipse-Dominator commented:
Upon some further investigation, AddCommand is also prone to having ignore parameter, as such it might be better to include the duplicated field as a separate parameter in the constructor rather than modifying the EditPersonDescriptor 

The new change shifts the warning message to Messages.
The code can further be abstracted for AddCommand and EditCommand to extend from parseCommand to reduce some level of code duplication but I feel that this might be too much of an overkill for this case.

Eclipse-Dominator committed:
Show warning when ignoring repeated parameters

Repeated occurrences of the same parameter in a command are being
ignored without any notification to the user.
This can lead to potential user errors going unnoticed, especially when
using the wrong parameter prefix multiple times.

Let's introduces a warning system that alerts the user when a parameter
is being ignored.
Now, if any parameter is repeated in a command, a warning message ensuring
that the user know which all the parameters are being duplicated.

damithc commented on src/main/java/seedu/address/logic/commands/AddCommand.java:
```diff
@@ -37,13 +42,23 @@ public class AddCommand extends Command {
+     * Creates an AddCommand to add the specified {@code Person} with duplicated fields {@code duplicateFields}.
      */
-    public AddCommand(Person person) {
+    public AddCommand(Person person, Set<Prefix> duplicateFields) {
```
I don't think this is a good approach. AddCommand should not need to know about duplicate fields.

Eclipse-Dominator replied:
Maybe we can make AddCommand throw exception for duplicate fields?
and make EditCommand show the relevant warning? 

That way we can also ground duplicate field to the EditPersonDescriptor

damithc replied:
A non-trivial design change is required to accommodate this warning message. The resulting design will be more complicated than the current one. Perhaps it is an acceptable cost if we want to optimize the application for a better user experience but AB3 also has to optimize for novice programmers. So, I think we may have to strip out this functionality altogether, for the sake of keeping the code tenable for novice programmers.

damithc reviewed (commented):
It looks like adding this warning is creating unnecessary complications. The alternative is to remove that feature altogether i.e., repeated parameters will be considered an error. Can you put this on hold for now and create a PR to remove that feature? After that, we can compare pros/cons of the two ways to add a warning and removing the feature entirely.

Eclipse-Dominator commented:
> It looks like adding this warning is creating unnecessary complications. The alternative is to remove that feature altogether i.e., repeated parameters will be considered an error. Can you put this on hold for now and create a PR to remove that feature? After that, we can compare pros/cons of the two ways to add a warning and removing the feature entirely.

See #190

Eclipse-Dominator commented:
Repeated Parameters show warning: #176 
Repeated Parameters show warning of repeated prefixes: #181 
Repeated Parametes treated as errors: #190

damithc commented:
Closing in favor of #190
